src/net/ossl.cc

```diff
@@ -1657,9 +1581,181 @@ class session : public enable_shared_from_this<session>, public session_impl {
+
+    if (BIO_get_init(b) <= 0 && ctrl != BIO_C_SET_POINTER) {
+        return 0;
+    }
+
+    auto session = unwrap_bio_ptr(b);
+
+    switch(ctrl) {
+    case BIO_C_SET_POINTER:
+        if (BIO_get_init(b) <= 0) {
+            BIO_set_data(b, data);
+            BIO_set_init(b, 1);
+        } else {
+            ret = 0;
+        }
+        break;
+    case BIO_CTRL_GET_CLOSE:
+        ret = BIO_get_shutdown(b);
+        break;
+    case BIO_CTRL_SET_CLOSE:
+        BIO_set_shutdown(b, static_cast<int>(num));
+        break;
+
+    case BIO_CTRL_DUP:
+    case BIO_CTRL_FLUSH:
+        ret = 1;
+        break;
+    case BIO_CTRL_EOF:
+        ret = BIO_test_flags(b, BIO_FLAGS_IN_EOF) != 0;
+        break;
+    case BIO_CTRL_PENDING:
+        ret = static_cast<long>(session->_input.size());
+        break;
+    case BIO_CTRL_WPENDING:
+        ret = session->_output_pending.available() ? 0 : 1;
+        break;
+    default:
+        ret = 0;
+    }
+    return ret;
+}
+
+/// This is called when the BIO is created
+///
+/// It is important for this to be set, even if it doesn't do anything
+/// because the BIO init flag won't get automatically set to '1'.
+int bio_create(BIO*) {
+    return 1;
+}
+
+/// Handles writes to the BIO
+///
+/// This function will attempt to call _out.put() and store the future in
+/// _output_pending.  If _output_pending has not yet resolved, return '0'
+/// and set the retry write flag.
+int bio_write_ex(BIO* b, const char * data, size_t dlen, size_t * written) {
+    auto session = unwrap_bio_ptr(b);
+    BIO_clear_retry_flags(b);
+
+    if (!session->_output_pending.available()) {
+        BIO_set_retry_write(b);
+        return 0;
+    }
+
+    try {
+        size_t n;
+
+        if (!session->_output_pending.failed()) {
+            scattered_message<char> msg;
+            msg.append(std::string_view(data, dlen));
+            n = msg.size();
+            session->_output_pending = session->_out.put(std::move(msg).release());
+        }
+
+        if (session->_output_pending.failed()) {
+            std::rethrow_exception(session->_output_pending.get_exception());
+        }
+
+        *written = n;
+        return 1;
+    } catch(const std::system_error & e) {
+        errno = e.code().value();
+        session->_output_pending = make_exception_future<>(std::current_exception());
+    } catch(...) {
+        errno = EIO;
+        session->_output_pending = make_exception_future<>(std::current_exception());
+    }
+    
+    return 0;
+}
+
+/// Handles reading data from the BIO
+///
+/// This will check to see if EOF has been reached and set the EOF
+/// flag if EOF has been reached.  It will set the retry read flag
+/// if no data is available, otherwise it will copy data off of
+/// the _input buffer and return it to the caller.
+int bio_read_ex(BIO* b, char * data, size_t dlen, size_t *readbytes) {
+    auto session = unwrap_bio_ptr(b);
+    BIO_clear_retry_flags(b);
+    if (session->eof()) {
```

Comment:
q: In case theres an exception set in the session, will eof be true or do additional checks need to be added here?